File: AoC2022/Day9.cs
using AoC.Shared;
using System;
namespace AoC2022
{
    public class Day9 : BaseDay
    {
        private HashSet<string> _visited;
        private Coordinate2d _positionHead;
        private Coordinate2d _positionTail;
        private Coordinate2d[] _segments;

        public Day9(int day, int year, bool isTest) : base(day, year, isTest) { }

        public Day9(int day, int year) : base(day, year) { }

        public override object SolvePart1()
        {
            _segments = Enumerable.Range(0, 2).Select(_ => new Coordinate2d(0, 0)).ToArray();
            _positionHead = _segments[0];
            _positionTail = _segments[1];
            _visited = new() { _positionTail.ToString() };

            foreach (var line in FileContent)
            {
                Move(line.Split(' ')[0], int.Parse(line.Split(' ')[^1]));
            }

            return _visited.Count;
        }

        public override object SolvePart2()
        {
            _segments = Enumerable.Range(0, 10).Select(_ => new Coordinate2d(0, 0)).ToArray();
            _positionHead = _segments[0];
            _positionTail = _segments[9];
            _visited = new() { _positionTail.ToString() };

            foreach (var line in FileContent)
            {
                Move(line.Split(' ')[0], int.Parse(line.Split(' ')[^1]));
            }

            return _visited.Count;
        }

        private void Move(string direction, int steps)
        {
            for (int i = 0; i < steps; i++)
            {
                _positionHead.Move(direction);

                for (var j = 1; j < _segments.Length; j++)
                {
                    _segments[j].Follow(_segments[j - 1]);
                }

                _visited.Add(_positionTail.ToString());
            }
        }
    }
}

File: AoC2022.python/day9.py
visited = set([(0, 0)])

H = [0, 0]
T = [0, 0]

for line in open('input/day9_test.txt'):
    x, y = line.split()
    y = int(y)
    for s in range(y):
        dx = 1 if x == "R" else -1 if x == "L" else 0
        dy = 1 if x == "U" else -1 if x == "D" else 0

        H[0] += dx
        H[1] += dy

        sx = H[0] - T[0]
        sy = H[1] - T[1]

        if (abs(sx) > 1 or abs(sy) > 1):
            if sx == 0:
                T[1] += sy // 2
            elif sy == 0:
                T[0] += sx // 2
            else:
                T[0] += 1 if sx > 0 else -1
                T[1] += 1 if sy > 0 else -1
        
        visited.add(tuple(T))
        
print(len(visited))
